### ISSUE TITLE
mergeAll does not accept arrays anymore

### ISSUE THREAD

**Issue by swemail** [ramda/types#22]
Hi and thank you for the ramda!

We just upgraded to 0.29.0 and the following code does not type check
```javascript
import {mergeAll} from 'ramda';

const list: Array<{x: number}> = [{x: 1}, {x: 2}];
const result = mergeAll(list);
```
It fails with `Argument of type '{ x: number; }[]' is not assignable to parameter of type '[{ x: number; }, ...{ x: number; }[]]'.
  Source provides no match for required element at position 0 in target.`

Have to do something like
```javascript
const list: Array<{x: number}> = [{x: 1}, {x: 2}];
const result = mergeAll([list[0], ...list]);
```
to make it work. The js code works with passing in arrays. I this intended?

**Comment by Harris-Miller** [comment]
Hi there thanks for the issue. I found out what was going on here. For `types-ramda` we updated to the latest `ts-toolkit` as that was needed to fix some existing types. The way the `mergeAll` was previously types used type functions from `ts-toolkit` that no longer exists. So we had to re-type them.

The current type is this:
```typescript
export function mergeAll<O extends object, Os extends readonly object[]>(list: [O, ...Os]): _.O.Assign<O, Os>;
```

`_.O.Assign` is defined [here](https://millsp.github.io/ts-toolbelt/modules/object_assign.html) and as the name suggests, a type for `Object.assign`, which is exactly what `mergeAll` is [under the hood](https://github.com/ramda/ramda/blob/v0.29.0/source/mergeAll.js#L25)

Our [tests](https://github.com/ramda/types/blob/develop/test/mergeAll.test.ts) included the following
```typescript
const foo = { foo: 'foo' };
const bar = { bar: 'bar' };
const foo2 = { foo: 2 };

expectType<{ foo: string; }>(mergeAll([foo]));
expectType<{ foo: string; bar: string; }>(mergeAll([foo, bar]));
expectType<{ foo: number; }>(mergeAll([foo, foo2]));
```

Those tests, as it turns out, only account for passing in array literals of either the same or different types. It turns out, that when you pass in an `Array<any>` or `any[]`, the typings break down. It breaks down specifically because when you have `Array<any>` or `any[]`, `[O, ...Os]` cannot be inferred in the same way when passing an array literal or tuple, Giving you the `Source provides no match for required element at position 0 in target.` error

Fortunately, this is an easy fix. Since when you `mergeAll` an `Array<T>` you simple get back `T`.  Which means the solution is just to add this overload

```typescript
export function mergeAll<T>(list: readonly T[]): T;
```

MR here: https://github.com/ramda/types/pull/23
I'll try to get this merged and released today for you
